pkg/controller/import-controller.go

```diff
@@ -44,315 +51,334 @@ const (
+	reconciler := &ImportReconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		CdiClient:  cdiClient,
+		K8sClient:  k8sClient,
+		Log:        log.WithName("import-controller"),
+		Image:      importerImage,
+		Verbose:    verbose,
+		PullPolicy: pullPolicy,
+		recorder:   mgr.GetRecorder("import-controller"),
 	}
-	return c
-}
-
-func (ic *ImportController) findScratchPvcFromCache(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
-	return ic.pvcLister.PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name + "-scratch")
-}
-
-func (ic *ImportController) findImportPodFromCache(pvc *v1.PersistentVolumeClaim) (*v1.Pod, error) {
-	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{LabelImportPvc: pvc.Name}})
+	importController, err := controller.New("import-controller", mgr, controller.Options{
+		Reconciler: reconciler,
+	})
 	if err != nil {
 		return nil, err
 	}
-
-	podList, err := ic.podLister.Pods(pvc.Namespace).List(selector)
-	if err != nil {
+	if err := addImportControllerWatches(mgr, importController); err != nil {
 		return nil, err
 	}
+	return importController, nil
+}
 
-	if len(podList) == 0 {
-		return nil, nil
-	} else if len(podList) > 1 {
-		return nil, errors.Errorf("multiple pods found for import PVC %s/%s", pvc.Namespace, pvc.Name)
+func addImportControllerWatches(mgr manager.Manager, importController controller.Controller) error {
+	// Add schemes.
+	if err := corev1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
 	}
-	return podList[0], nil
-}
 
-func (ic *ImportController) requiresScratchSpace(pvc *v1.PersistentVolumeClaim) bool {
-	scratchRequired := false
-	contentType := getContentType(pvc)
-	// All archive requires scratch space.
-	if contentType == "archive" {
-		scratchRequired = true
-	} else {
-		switch getSource(pvc) {
-		case SourceGlance:
-			scratchRequired = true
-		case SourceRegistry:
-			scratchRequired = true
-		}
+	// Setup watches
+	if err := importController.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return isPVCInteresting(e.Object.(*corev1.PersistentVolumeClaim))
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return isPVCInteresting(e.ObjectNew.(*corev1.PersistentVolumeClaim))
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return isPVCInteresting(e.Object.(*corev1.PersistentVolumeClaim))
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}); err != nil {
+		return err
 	}
-	value, ok := pvc.Annotations[AnnRequiresScratch]
-	if ok {
-		boolVal, _ := strconv.ParseBool(value)
-		scratchRequired = scratchRequired || boolVal
+	if err := importController.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
```

Comment:
The PVCs that own these pods may not be "interesting"

Reply:
So as far as I could tell, something that triggers the Pod  watch, which is then translated to a PVC, then goes through the PVC watch, so it will do the check. Maybe I should rename isPVCInteresting, to shouldReconcilePVC?

Reply:
But I will verify this is the case

Reply:
You are right, it bypasses the check. I added another check in the reconcile loop

Reply:
Now it is checked in both places.  Maybe it is clearer to just do it in the reconcile loop?

Also, I'm not sure that the check that was added to the reconcile loop is in the right place.  Why is it not called right after getting the PVC, like maybe here:  https://github.com/kubevirt/containerized-data-importer/blob/c9b741974233f120f8eef35649cbe75c96b334fc/pkg/controller/import-controller.go#L158

Reply:
Let me take another look at what I am doing in the reconcile loop, there is some trickiness when I delete a pod triggering a reconcile and causing it to create a new importer pod which results in an endless loop of create/delete pods. But that was happening before I did the check. So I had to delete in the next reconcile loop, but if I have the check I can do both in the same loop.

Reply:
So the problem turned out to be in checkPVC, it would return false if the termination was not nil, which works in most cases, except where the PVC is removed externally and then it fails. So once fixed, I was to make a clean reconcile loop.